e2core/server/handlers.go

```diff
@@ -1,119 +1,131 @@
+				// header names, which ours are.
+				if c.Response().Header()[head] == nil {
+					c.Response().Header()[head] = make([]string, 0)
+				}
+
+				c.Response().Header()[head] = append(c.Response().Header()[head], val)
 			}
 		}
 
 		responseData := seq.Request().State[mod.FQMN]
 
-		return vk.RespondBytes(ctx.Context, w, responseData, http.StatusOK)
+		return c.Blob(http.StatusOK, "text/plain", responseData)
 	}
 }
 
-func (s *Server) executePluginByRefHandler() vk.HandlerFunc {
-	return func(w http.ResponseWriter, r *http.Request, ctx *vk.Ctx) error {
-		ref := readParam(ctx, "ref")
+func (s *Server) executePluginByRefHandler(l zerolog.Logger) echo.HandlerFunc {
+	ll := l.With().Str("handler", "executePluginByRefHandler").Logger()
+
+	return func(c echo.Context) error {
+		ref := c.Param("ref")
 
 		mod := s.syncer.GetModuleByRef(ref)
 		if mod == nil {
-			return vk.E(http.StatusNotFound, "module not found")
+			return echo.NewHTTPError(http.StatusNotFound, "module not found")
 		}
 
-		ctx.Log.Debug("found module by ref:", mod.FQMN)
+		ll.Debug().Str("fqmn", mod.FQMN).Msg("found module by ref")
 
-		req, err := request.FromVKRequest(r, ctx)
+		req, err := request.FromEchoContext(c)
 		if err != nil {
-			ctx.Log.Error(errors.Wrap(err, "failed to request.FromVKRequest"))
-			return vk.E(http.StatusInternalServerError, "failed to handle request")
+			return echo.NewHTTPError(http.StatusInternalServerError).SetInternal(err)
 		}
 
-		req.UseSuborbitalHeaders(r, ctx)
+		err = req.UseSuborbitalHeaders(c)
+		if err != nil {
+			return err
+		}
 
 		steps := []tenant.WorkflowStep{{FQMN: mod.FQMN}}
 
 		// a sequence executes the handler's steps and manages its state.
 		seq, err := sequence.New(steps, req)
 		if err != nil {
-			ctx.Log.Error(errors.Wrap(err, "failed to sequence.New"))
-			return vk.E(http.StatusInternalServerError, "failed to handle request")
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to handle request")
 		}
 
 		if err := s.dispatcher.Execute(seq); err != nil {
-			ctx.Log.Error(errors.Wrap(err, "failed to Execute"))
-			return vk.E(http.StatusInternalServerError, "failed to execute plugin")
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to execute plugin")
 		}
 
 		// handle any response headers that were set by the Runnables.
 		if req.RespHeaders != nil {
 			for head, val := range req.RespHeaders {
-				w.Header().Add(head, val)
+				// need to directly assign because .Add and .Set will filter out non-standard
+				// header names, which ours are.
+				if c.Response().Header()[head] == nil {
+					c.Response().Header()[head] = make([]string, 0)
+				}
+
+				c.Response().Header()[head] = append(c.Response().Header()[head], val)
 			}
 		}
 
 		responseData := seq.Request().State[mod.FQMN]
 
-		return vk.RespondBytes(ctx.Context, w, responseData, http.StatusOK)
+		return c.Blob(http.StatusOK, "text/plain", responseData)
```

Comment:
It's very odd to me that we're returning something here when the return type is an error (even though I assume c.Blob returns nil for successful responses)

Is this an Echo norm?

Reply:
Also text/plain seems wrong here... I would think that it'd be reasonable to default to `application/json`, wdyt?

Reply:
I looked at what `vk.RespondBytes` did, then looked at what the content type should be for random things. According to MDN, if it's human readable text, it should be `text/plain`, otherwise `application/octet-stream` for raw binary data, or if it's any of the files, then that.

`application/json` only makes sense if the data being returned is actually a json string, which I don't think it is in this case.

The way responses work in echo, and gin, and fiber, and most of them that use a context is to use response methods, like `c.JSON`, `c.String`, `c.Blob`, etc, which will write to the http writer, and return an error if something went wrong. If everything goes well, and the response was sent to the client successfully, it returns nil, and the error handler doesn't need to kick in.

In a way it's an echo norm, but also a wider norm of "this is how web frameworks work".

Reply:
I would say `application/octet-stream` would make more sense than text-plain, personally.

Reply:
Changed it in https://github.com/suborbital/e2core/pull/365/commits/a5160b0b5892793c4d4ae1af31ebc012020e1b33